### PR TITLE
Basic name poisoning support

### PR DESCRIPTION
**Pull request by bricknerb**
https://github.com/carbon-language/carbon-lang/issues/4622
When using an unqualified name, disallow declaring that name in all scopes that would make it ambiguous in retrospect.
Doesn't include support for poisoning in `impl library` (see new test for that with TODO).
Implemented by introduce `InstId::PoisonedName` and entries with it to `NameScope`.

### CODE REVIEW

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (jonmeow) [inline toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:246]: Is a name reference to the poisoned entity tested? e.g.:

```suggestion

// --- fail_use_poison.carbon

library "[[@TEST_NAME]]";

class C {}

namespace N;
fn N.F() -> C;

fn F() -> N.C;

```

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (jonmeow) [inline toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon]:
```suggestion
// --- fail_poison_import.carbon
```

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (jonmeow) [inline toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:250]:
```suggestion
library "[[@TEST_NAME]]";
import library "poison";
```

Should declare the library name

File: toolchain/sem_ir/name_scope.h
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.h]: Generally I'd suggest writing comments like you would markdown:

- Put a blank line inbetween if you want separate paragraphs. i.e.:

```suggestion
  // not.
  //
  // If the name is found, it might be a poisoned name.
```

- Wrap if you want a single paragraph:

```suggestion
  // not. If the name is found, it might be a poisoned name.
```

- Use bullets (`// - text here`) if you're trying to list things. (probably doesn't apply in this comment)

Just noting this for style of below comments too.

File: toolchain/sem_ir/name_scope.cpp
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.cpp]: Should this check before the insert, so that it fails even if the name was previously inserted?

File: toolchain/sem_ir/name_scope.cpp
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.cpp]:
```suggestion
               "Trying to poison an existing non-poisoned name: {0}", name_id);
```

File: toolchain/check/context.cpp
Comment (jonmeow) [inline toolchain/check/context.cpp]: We prefer LLVM types over std types in most cases.

```suggestion
  llvm::SmallVector<LookupScope> scopes_to_poison;
```

File: toolchain/check/context.cpp
Comment (jonmeow) [inline toolchain/check/context.cpp]: Had you considered not saving scopes where `non_lexical_result.inst_id.is_poisoned()` is true, in order to avoid poisoning twice? Would that allow `AddPoison` to error on _all_ insert failures, not just non-poisoned ones?

File: toolchain/sem_ir/ids.h
Comment (jonmeow) [inline toolchain/sem_ir/ids.h]:
```suggestion
    return *this == PoisonedName;
```

File: toolchain/sem_ir/ids.h
Comment (jonmeow) [inline toolchain/sem_ir/ids.h]:
```suggestion
constexpr InstId InstId::PoisonedName = InstId(InvalidIndex - 1);
```

File: toolchain/sem_ir/ids.h
Comment (jonmeow) [inline toolchain/sem_ir/ids.h]:
```suggestion
```

For style with [declaration order](https://google.github.io/styleguide/cppguide.html#Declaration_Order), this static constant would belong at the top. However, I'm suggesting a few changes to remove this, similar to what we have elsewhere in this file.

File: toolchain/check/decl_name_stack.cpp
Comment (jonmeow) [inline toolchain/check/decl_name_stack.cpp]:
```suggestion
    case DeclNameStack::NameContext::State::Poisoned:
    case DeclNameStack::NameContext::State::Unresolved:
```

Very minor, but when there's no particular significance to the ordering, I'll suggest keeping things in lexical order.

File: toolchain/check/handle_class.cpp
Comment (jonmeow) [inline toolchain/check/handle_class.cpp]:
```suggestion
    // This is a declaration of a poisoned name.
```

File: toolchain/check/context.cpp
Comment (jonmeow) [inline toolchain/check/context.cpp]: From this diagnostic, I'm thinking that we should take a different approach that allows us to annotate the location which introduced the poisoned name. i.e., something like:

```
// error: the NameDeclPoisoned error
// class N.C;
//       ^~~
// note: unqualified name lookup here
// fn N.F() -> C;
//             ^
```

That might require changing the way poisoned entries are tracked, to keep the declaration that _caused_ the poisoning with the `NameScope::Entry`. This is having me wonder if we should have something like:

```
  struct Entry {
    NameId name_id;
    // Maybe this should actually be a `union` with a `LocId` based on `is_poisoned`?
    InstId inst_id;
    AccessKind access_kind;
    bool is_poisoned;
  };
  static_assert(sizeof(Entry) == 12);
```

(noting the `static_assert` to emphasize packing)

The reason I'm thinking in that direction is because I'm now thinking we should actually use `inst_id` here to point back to the instruction which introduced the poisoning (there may be a better way of doing that, too).

Sorry, I know I'd commented in favor of the special `InstId`. I think though that this could be done as an incremental change on top of what you have here. I'd be fine if you want to leave something like this as a TODO or for a separate PR. I just wanted to note the thought that we may need to change approaches for a better diagnostic. Let me know how you'd like to proceed here.

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (jonmeow) [inline toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon]: As discussed, this shouldn't fail because it would be cross-library. However, it should fail for an `impl library`, such as:

```
// --- fail_poison.impl.carbon

// "TEST_NAME" here translates to "poison", which is what you'd want.
impl library "[[@TEST_NAME]]";

class N.C {}
```

File: toolchain/check/context.cpp
Comment (jonmeow) [inline toolchain/check/context.cpp]:
```suggestion
      "name previously used by unqualified name lookup and not found; cannot later be declared");
```

**Review by jonmeow — COMMENTED** [review]
Looks good! This made me think a little more about what's needed to get a good diagnostic -- some comments inline, sorry I didn't think about that before.

File: toolchain/check/context.cpp
Comment (jonmeow) [reply toolchain/check/context.cpp]: We have a little [diagnostic style guide](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide). Generally I think we're trying to focus on the situation the toolchain observed first. Suggesting a rephrasing which I _think_ better reflects the intent.

Comment (zygoloid) [reply toolchain/check/context.cpp]: I wonder if it'd be better to put the diagnostic at the point of use, and say something like "name used before it was declared" with a note pointing to the later declaration?

Comment (jonmeow) [reply toolchain/check/context.cpp]: That seems reasonable, though noting [my comments about location above](https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876530039), so for now this would probably look like using an `Invalid` ID for the location that needs to be piped through:

```
CARBON_DIAGNOSTIC(NameUseBeforeDecl, Error, "name used before it was declared");
CARBON_DIAGNOSTIC(NameUseBeforeDeclNote, Note, "declared here");
// TODO: This should pass in the original declaration.
unit_info.emitter.Emit(Parse::NodeId::Invalid, NameUseBeforeDecl)
    .Note(decl_loc, NameUseBeforeDeclNote)
    .Emit();
```

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Done.

File: toolchain/check/decl_name_stack.cpp
Comment (bricknerb) [reply toolchain/check/decl_name_stack.cpp]: Done.

File: toolchain/check/handle_class.cpp
Comment (bricknerb) [reply toolchain/check/handle_class.cpp]: Done.

File: toolchain/sem_ir/ids.h
Comment (bricknerb) [reply toolchain/sem_ir/ids.h]: Done.

File: toolchain/sem_ir/ids.h
Comment (bricknerb) [reply toolchain/sem_ir/ids.h]: Done.

File: toolchain/sem_ir/ids.h
Comment (bricknerb) [reply toolchain/sem_ir/ids.h]: Done.

File: toolchain/sem_ir/name_scope.cpp
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.cpp]: Done.

File: toolchain/sem_ir/name_scope.h
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.h]: Done.

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (bricknerb) [reply toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon]: Done.

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (bricknerb) [reply toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:250]: Done.

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (bricknerb) [reply toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon]: Done.

File: toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon
Comment (bricknerb) [reply toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:246]: Done, though this was implicitly tested in fail_poison_with_usage.

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Thanks!
I've changed the diagnostic name and text.
It makes sense to put the warning where the name is poisoned, though I find an error in code that is caused because of code that comes later somewhat confusing.
I'm ok with both options.
Anyways, this needs to wait for when we store the data on what caused the poisoning.

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Done.

File: toolchain/sem_ir/name_scope.cpp
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.cpp]: Done.

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Thanks Jon!
Yes, this was my concern with this option in https://docs.google.com/document/d/16vhcVeZN2w8iIl_a6gNxd5kFZPPfWh3f5_TnaK9sdHQ

> Since it uses the existing data structure, it’s somewhat complex to add specific metadata for diagnosis. For example, adding a reference to where the name was poisoned doesn’t fit well into this design.

I thought of manipulating `Entry` for that but there is some concern that it makes the structure more complex to use and extend, compared to having a separate set, which might make the lookup more complex.
Since this might change this PR significantly, I think I'll make that a TODO so we can have name poisoning support and improve the diagnostics in a separate PR as suggested.
Added a TODO in `Context::DiagnosePoisonedName()`.

Comment (jonmeow) [reply toolchain/check/context.cpp]: I think we're likely to prefer the structure complexity versus a new set because it's likely higher performance.

File: toolchain/check/context.cpp
Comment (jonmeow) [reply toolchain/check/context.cpp]: Unresolving -- I think we should probably go with the phrasing zygoloid suggested, with TODOs per https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876772138

File: toolchain/sem_ir/name_scope.h
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.h]:
```suggestion
  // Adds a name that is required to exist in a name scope, such as `Self`.
  // The name must never conflict. inst_id must not be poisoned.
```

File: toolchain/sem_ir/name_scope.h
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.h]:
```suggestion
  // If the given name already exists, return true with the EntryId; the entry
  // might be poisoned. Otherwise, adds the name using inst_id and access_kind
  // and returns false with the new EntryId.
  //
  // This cannot be used to add poisoned entries; use AddPoison instead.
```

Suggesting a slightly different phrasing here, since I'm suggesting the paragraph edit too.

File: toolchain/sem_ir/name_scope.h
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.h]:
```suggestion
  // Searches for the given name and returns an EntryId if found or nullopt if
  // not. The returned entry may be poisoned.
```

I'm going to suggest using fewer paragraphs. The context and briefer read can often be helpful, and paragraphs are typically a few sentences anyways.

File: toolchain/sem_ir/name_scope.h
Comment (jonmeow) [inline toolchain/sem_ir/name_scope.h]:
```suggestion
  // Adds a new name known to not exist. Must not be poisoned.
```

File: toolchain/sem_ir/name_scope.h
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.h]: Done.

File: toolchain/sem_ir/name_scope.h
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.h]: Done.

File: toolchain/sem_ir/name_scope.h
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.h]: Done.

File: toolchain/sem_ir/name_scope.h
Comment (bricknerb) [reply toolchain/sem_ir/name_scope.h]: Done.

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Right now we have the location of the later declaration.
IIUC, the suggestion is to error where the poisoning happened with a note on the later declaration.
If we go with this suggestion, we only have the note, not the error.
Can you clarify what you suggest for this PR?
I've made the TODO clearer.

File: toolchain/check/context.cpp
Comment (jonmeow) [inline toolchain/check/context.cpp]:
```suggestion
  CARBON_DIAGNOSTIC(NameUseBeforeDecl, Error, "name used before it was declared");
  CARBON_DIAGNOSTIC(NameUseBeforeDeclNote, Note, "declared here");
  unit_info.emitter.Emit(Parse::NodeId::Invalid, NameUseBeforeDecl)
      .Note(loc, NameUseBeforeDeclNote)
      .Emit();
```

File: toolchain/check/context.cpp
Comment (jonmeow) [inline toolchain/check/context.cpp]:
```suggestion
  // TODO: Improve the diagnostic to replace NodeId::Invalid with the location where the name was poisoned. See discussion in
```

File: toolchain/check/context.cpp
Comment (jonmeow) [reply toolchain/check/context.cpp]: Sure, I've copied in my code from the above comment into a suggested edit: https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1880640674.

Comment (jonmeow) [reply toolchain/check/context.cpp]: Can you elaborate why you think there will only be the note and not the error? What does the delta to tests look like?

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Done.

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Thanks! Done.
I was not aware that it's ok (for now) to put an invalid location, and that's why I thought we won't be able to error if we don't use the location we have.

File: toolchain/check/context.cpp
Comment (bricknerb) [reply toolchain/check/context.cpp]: Thanks for the clarification.
Done.
See my reply in the other comment.

**Comment by jonmeow** [comment]
Can you take a look at the test failures?

**Comment by bricknerb** [comment]
> Can you take a look at the test failures?

Sure, I couldn't reproduce this locally and couldn't find a diff between what I run locally and the failure.
I obviously do something wrong here.
Do you have any pointers how can I reproduce this?

**Comment by zygoloid** [comment]
It's probably only failing after merging with trunk.

**Comment by bricknerb** [comment]
> It's probably only failing after merging with trunk.

Thanks!
This makes sense, just wasn't aware this is what it does.
I think I've fixed it, but I can probably not rerun it myself.

**Comment by jonmeow** [comment]
I see this failed on another merge conflict. I've merged to try to get it in now. The contributor access should help avoid this repeating, though; I encourage resolving conflicts & merging after approval, now that you can. :)

**Review by jonmeow — APPROVED** [review]

**Comment by jonmeow** [comment]
And maybe I just missed that I hadn't approved...

**Comment by bricknerb** [comment]
Thanks for the merge Jon!
Indeed I was waiting for an approval. :)

**Pull request merged**
